tests/integration/operator/operator_integration_test.go

```diff
@@ -0,0 +1,97 @@
+package operator
+
+import (
+	"testing"
+
+	"github.com/gruntwork-io/terratest/modules/helm"
+	"github.com/gruntwork-io/terratest/modules/k8s"
+	. "github.com/k8ssandra/k8ssandra/tests/integration/util"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var (
+	suiteName           = "k8ssandra-integration-suite"
+	operatorReleaseName = "k8ssandra"
+	namespace           = "k8ssandra"
+	clusterReleaseName  = "cassdc"
+	operatorChart       = "../../../charts/k8ssandra"
+	clusterChart        = "../../../charts/k8ssandra-cluster"
+)
+
+func Test(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, suiteName)
+}
+
+// setup used for creating test option context and cleanup of operator and cluster.
+func setup() K8ssandraOptionsContext {
+
+	kubeOptions := k8s.NewKubectlOptions("", "", namespace)
+	operatorOptions := &helm.Options{KubectlOptions: kubeOptions}
+	clusterOptions := &helm.Options{KubectlOptions: kubeOptions}
+
+	ctx := CreateK8ssandraOptionsContext(operatorOptions, clusterOptions)
+	UninstallRelease(ctx.Cluster, clusterReleaseName)
+	UninstallRelease(ctx.Operator, operatorReleaseName)
+	return ctx
+}
+
+var _ = Describe(suiteName, func() {
+
+	var options K8ssandraOptionsContext
+
+	Describe("baseline required preconditions", func() {
+
+		It("should perform setup followed by operator install", func() {
+
+			options = setup()
+
+			By("Expecting operator and cluster options to be setup")
+			Ω(options).ShouldNot(BeNil())
+			Ω(options.Cluster).ShouldNot(BeNil())
+			Ω(options.Operator).ShouldNot(BeNil())
+
+			InstallChart(options.Operator, operatorChart, operatorReleaseName, "")
+
+			By("Expecting the operator to be in a deployed status")
+			Ω(IsReleaseDeployed(options.Operator, "k8ssandra")).Should(BeTrue())
+		})
+	})
+
+	Describe("k8ssandra cluster integration", func() {
+
+		BeforeEach(func() {
+			options = setup()
+		})
+
+		It("should install utilizing charts; k8ssandra operator then cluster", func() {
+
+			InstallChart(options.Operator, operatorChart, operatorReleaseName, "")
+			InstallChart(options.Cluster, clusterChart, clusterReleaseName, "")
+
+			By("Expecting to have the operator and cluster in deployed status")
+			Ω(IsReleaseDeployed(options.Operator, "k8ssandra")).Should(BeTrue())
+			Ω(IsReleaseDeployed(options.Cluster, "cassdc")).Should(BeTrue())
+
+			By("Expecting to have running cass-operator pod")
+			Ω(IsPodRunning(options.Operator, "cass-operator")).Should(BeTrue())
```

Comment:
Not sure I understand the purpose of the `IsPodRunning` function. It looks like it just checks to see if there are _any_ running pods in the specified namespace. The first and most important thing we want to check is that the CassandraDatacenter has been created and is ready.

Reply:
@jsanda 
Yes, really happy to get this one behind us, so feel free to expand a bit if the below doesn't assist with the intention of this test operation.

---

Working backwards a bit, there is a `LookupRunningPods` function in the `test util module` that performs lookup of only running pods scoped by namespaces.  Many of the util functions are there to introduce "building-blocks" for reuse.

https://github.com/k8ssandra/k8ssandra/blob/725090f0fa107586a105ca279876e7114704314f/tests/integration/util/lib.go#L154

```go
result, err := k8s.RunKubectlAndGetOutputE(GinkgoT(), ctx.kubeOptions, "get", "pods", "-n",
		ctx.namespace, "--field-selector=status.phase=Running", "-o", "name", "--no-headers=true")
```
The `IsPodRunning` function invokes the `LookupRunningPods` with a prefix name and in this case, identifies our 'cass-operator'.   It returns a true status when the prefix name is located in that set.  False otherwise.

Example:
`cass-operator-cd9b57568-mw596                        1/1     Running   2          14h`

https://github.com/k8ssandra/k8ssandra/blob/725090f0fa107586a105ca279876e7114704314f/tests/integration/util/lib.go#L121

There is also a verification of running pods at the label level for **cass-operator** as part of the util module.
```go
Ω(IsLabeledPodExisting(options.Operator, "cass-operator")).Should(BeTrue())Ω(IsLabeledPodExisting(options.Operator, "cass-operator")).Should(BeTrue())
```

This test is checking the running pods and the deployed releases specific to the deployed name `cass-operator` which is the CRD kind of `CassandraDataCenter`.

Reply:
>Many of the util functions are there to introduce "building-blocks" for reuse.

I can appreciate the bottom up approach of trying to put some infrastructure in place before we get too far ahead with tests; however, I think we would be better served right now with more of a top down approach with getting more test coverage in place first and then follow up with refactoring common code into shared packages/libraries. 

We have a spent a decent amount of time on this PR, but we still have not gotten to the critical verification step - check that the CassandraDatacenter is ready. 

With respect to the `IsPodRunning` and `LookupRunningPods` functions, they could be combined, e.g., 

```
$ kubectl get pods --field-selector=status.phase=Running -l name=cass-operator 
```

With a lot of these checks, you need to introduce some polling since things are done asynchronously. 

Lastly, we already have code in reaper-operator (that was also used in an old version of medusa-operator) for verifying that a Deployment is ready. Take a look [here](http://github.com/k8ssandra/reaper-operator/blob/d4834fcfac4d901f2269c66f57ed5a0d0b3d17c0/test/framework/framework.go#L128-L128). I am not saying that the code in reaper-operator has to be _the_ approach, but it certainly good to be aware of it and understand it.

Reply:
> With a lot of these checks, you need to introduce some polling since things are done asynchronously.

Polling and waits are built-in to the module functions: [wait](https://github.com/k8ssandra/k8ssandra/blob/a2e592fed685583a3849222da640f93cfc46a3a3/tests/integration/util/lib.go#L83) , [pause](https://github.com/k8ssandra/k8ssandra/blob/a2e592fed685583a3849222da640f93cfc46a3a3/tests/integration/util/lib.go#L93) , [pause-impl](https://github.com/k8ssandra/k8ssandra/blob/a2e592fed685583a3849222da640f93cfc46a3a3/tests/integration/util/lib.go#L190)